### PR TITLE
Use fullscreen triangles for fullscreen passes.

### PR DESCRIPTION
**Pull request by ACEfanatic02**
Instead of using a vertex buffer with a quad triangle strip, generate a single fullscreen triangle in the vertex shader.  This has a number of advantages:

- Avoids fragment quad overshading along the diagonal.

- Avoids creating, filling, reading, and destroying a vertex buffer.

### CODE REVIEW

**Comment by ZehMatt** [comment]
Do you have any numbers to how it affects it?

**Comment by janisozaur** [comment]
I want to test it with my raspberry pi

**Comment by ACEfanatic02** [comment]
> Do you have any numbers to how it affects it?

No.  The game currently crashes under DLL injection (see Discord for a discussion), which makes it difficult to profile externally, and explicit timing queries have not yet been added.  With the (lack of) complexity of these fragment shaders, I'd be shocked if there was a measurable GPU performance difference in either direction, though.  The main win here in practice is less code and fewer allocations to track.

**Comment by ZehMatt** [comment]
Also gonna tag @LRFLEW, he is the original author for most of the code

**Comment by LRFLEW** [comment]
Man, it's been a while since I've been active in this project, but thank you for the ping.

Yes, the code for ApplyPaletteShader and ApplyTransparencyShader were my contributions as part of #6541. However, a lot of that code was based on even older code for the now defunct CopyFramebufferShader (which I replaced with `glBlitFramebuffer`).

Using the fullscreen triangle makes sense. It's not something I was taught (so wouldn't have thought of back in the day), but looking into the idea, [it does seem legit](https://wallisc.github.io/rendering/2021/04/18/Fullscreen-Pass.html), so I support that change.

It's the other part of this PR I'm unsure of. Pretty much all the vertex shaders used to use an index like this PR uses, though [implemented in a more complicated way](https://github.com/OpenRCT2/OpenRCT2/blob/85b257cc2e216f8a664a079739aad26d610dfe26/data/shaders/copyframebuffer.vert#L14-L32) and [still used a VBO](https://github.com/OpenRCT2/OpenRCT2/blob/85b257cc2e216f8a664a079739aad26d610dfe26/src/openrct2-ui/drawing/engines/opengl/CopyFramebufferShader.cpp#L25-L34). I opted to replace that system with a simpler VBO-based system in my rewrite, as it seemed like an improvement.

> Avoids creating, filling, reading, and destroying a vertex buffer.

The creation, filling, and destruction of the VBO is only done once per process, so I don't see that as a major performance concern. Similarly, the data is only 16 floats, so I don't think it's a significant VRAM consideration. The only question, then, is whether having the GPU read the VBO is significantly slower than computing the values in the shader. I would assume not, since GPUs are optimized for processing VBO and Texture data, and having only 3 vertices reduces the GPU's ability to parallelize the computation. In terms of code quality, I guess the question is whether we'd rather more complexity in the shader or more complexity in the supporting C++ code. Currently, I'm not sold on the change, and would like to see some numbers of possible on this.

**Comment by ACEfanatic02** [comment]
> The creation, filling, and destruction of the VBO is only done once per process, so I don't see that as a major performance concern. 

Removing the buffer isn't really about raw performance, but complexity reduction.  Code that doesn't exist can't have bugs.

> The only question, then, is whether having the GPU read the VBO is significantly slower than computing the values in the shader. I would assume not, since GPUs are optimized for processing VBO and Texture data, and having only 3 vertices reduces the GPU's ability to parallelize the computation.

There's two ways to look at this:
-  In isolation, looking at only the vertex shader, yes this will be faster.  Three integer ops, two conversions, and two FMAs will be faster than loading a cache line to read the vertex buffer.  The number of vertices is actually irrelevant in this case, because shaders run in a wide (32- or 64-thread) SIMT model.  Rendering one triangle versus two takes exactly the same number of VS threadgroups.
- In practice, this will not be faster _or_ slower, because there's no hazard with this vertex shader and so it will (most likely) spawn its one threadgroup in amongst the tail end of the previous pass's FS work where there are unoccupied shader units.  So either way the VS for fullscreen passes tends to be "free" unless it has an actual dependency (or if there is a manual barrier in place.)

On its own this is not a particularly inspiring change, I agree.  I wanted to start with something small to learn the PR / review process.

**Comment by ZehMatt** [comment]
We did a bit of house cleaning, this needs to rebased. I looked at the changes and I think they seem fine, when I had a brief glance initially I thought its doing something different. Should also wait for @LRFLEW, but I'm fine with this.

**Comment by LRFLEW** [comment]
I've left review comments on things I'd like to see changed in the shaders before this gets merged. If those are all addressed, I have no qualms with this being merged.

If we're using the VertexID trick here, it might be worth considering it for the other shaders, which use a small VBO in conjunction with instancing for drawing. It doesn't have to be with this PR, but I did just want to make a note of that.

**Review by LRFLEW — CHANGES_REQUESTED** [review]

**Review by janisozaur — APPROVED** [review]
It's the same as my baseline on ARM SBC

**Comment by Gymnasiast** [comment]
@ACEfanatic02 Could you address the line notes that @LRFLEW added?

**Comment by 733737** [comment]
keep open

**Comment by Gymnasiast** [comment]
@733737 I don’t think there is any point, we have been waiting for OP to address feedback given on 11 March, almost three months ago. I’m closing this.

@ACEfanatic02 Let us know if you intend to address this and we’ll reopen it.

**Pull request closed without merging**
